scripts/migration/add_timestamp_and_originator_to_transfers/main.go

```diff
@@ -0,0 +1,313 @@
+
+func main() {
+	cfg := new(config)
+	err := validatorCfg.GetConfig(cfg, "scripts/migration/add_timestamp_and_originator_to_transfers/config.yml")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Setup hedera, evm clients and nodes to be migrated
+	nodes := make([]*node, len(cfg.DBs))
+	for i, db := range cfg.DBs {
+		pgDb := persistence.NewPgConnector(db).Connect()
+		db, err := pgDb.DB()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		nodes[i] = &node{
+			db: db,
+		}
+	}
+	evmClients := make(map[uint64]client.EVM)
+	for k, v := range cfg.Evm {
+		evmClients[k] = evm.NewClient(validatorCfg.Evm{
+			NodeUrl:            v,
+			BlockConfirmations: 5,
+		}, k)
+	}
+
+	mnc := mirror_node.NewClient(cfg.Hedera.MirrorNode)
+
+	migrator := newMigrator(nodes, evmClients, mnc)
+
+	log.Infof("finna update %d nodes using %d evm clients", len(migrator.nodes), len(migrator.evmClients))
+	err = migrator.migrate()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+type tempTransfer struct {
+	TransactionID string
+	SourceChainID uint64
+	Timestamp     sql.NullTime
+	Originator    sql.NullString
+}
+
+type hederaCfg struct {
+	Env        hedera.NetworkName      `yaml:"env"`
+	AccountId  string                  `yaml:"account_id"`
+	PrivateKey string                  `yaml:"private_key"`
+	MirrorNode validatorCfg.MirrorNode `yaml:"mirror_node"`
+}
+
+type config struct {
+	Hedera hederaCfg               `yaml:"hedera"`
+	DBs    []validatorCfg.Database `yaml:"dbs"`
+	Evm    map[uint64]string       `yaml:"evm"`
+}
+
+type node struct {
+	db *sql.DB
+}
+
+type migrator struct {
+	nodes        []*node
+	evmClients   map[uint64]client.EVM
+	hederaClient client.MirrorNode
+}
+
+func newMigrator(nodes []*node, evmClients map[uint64]client.EVM, hederaClient client.MirrorNode) *migrator {
+	return &migrator{nodes: nodes, evmClients: evmClients, hederaClient: hederaClient}
+}
+
+func (m *migrator) migrate() error {
+	for i, _ := range m.nodes {
+		err := m.migrateNode(i)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *migrator) migrateNode(i int) error {
+	migratedNode := m.nodes[i]
+	c, err := migratedNode.db.Query(`select count (*) from transfers`)
+	if err != nil {
+		return err
+	}
+
+	var count int
+	if c.Next() {
+		if err := c.Scan(&count); err != nil {
+			return err
+		}
+	} else {
+		return errors.New("no transfers count")
+	}
+	c.Close()
+
+	tx, err := migratedNode.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	page := 1
+	const perPage = 50
+	totalUpdated := 0
+	for {
+		if totalUpdated >= count {
+			if err := tx.Commit(); err != nil {
+				return err
+			}
+			log.Infof("finished migrating %d transfers", totalUpdated)
+			break
+		}
+
+		transfers, err := migratedNode.pagedTransfers(page, perPage)
+		if err != nil {
+			return err
+		}
+
+		for _, t := range transfers {
+			if t.Originator.Valid && t.Timestamp.Valid {
+				continue
+			}
+
+			err := m.fillTimestampAndOriginator(t)
```

Comment:
Can we add logs for each transfer that we update - tx id, and the to-be-set originator & timestamp?

Reply:
Plus the DB we update it to as well